File: Minimum_Depth_of_Binary_Tree.cpp
 //
//Created on Feb 08, 2014
//
//@author: Xufan Du
//

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode *root) {
        if(root==NULL) return 0;
        if(root->left&&root->right) return 1+std::min(minDepth(root->left), minDepth(root->right));
        
        if(root->left) return 1+minDepth(root->left);
        
        return 1+minDepth(root->right);
    }
};

File: LeetCode/111. Minimum Depth of Binary Tree/minimum_depth_of_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def minDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        if not root.left:
            return self.minDepth(root.right) + 1
        elif not root.right:
            return self.minDepth(root.left) + 1
        else:            
            return min(self.minDepth(root.left), self.minDepth(root.right)) + 1
        
        
        
#notice the defination of 'leaf' node
#have to check if the node is 'leaf' node or not
